core/process.py

```diff
@@ -969,20 +995,42 @@ def WhenDone(self, pid, status):
     self.state = job_state_e.Done
     if self.parent_pipeline:
       self.parent_pipeline.WhenDone(pid, status)
+    else:
+      self.job_state.MaybeTakeTerminal()
 
   def RunProcess(self, waiter, why):
     # type: (Waiter, trace_t) -> int
     """Run this process synchronously."""
     self.StartProcess(why)
+    # ShellExecutor might be calling this for the last part of a pipeline.
+    if self.parent_pipeline is None:
+      # QUESTION: Can the PGID of a single process just be the PID?  i.e. avoid
+      # calling getpgid()?
+      self.job_state.MaybeGiveTerminal(posix.getpgid(self.pid))
     return self.Wait(waiter)
 
 
 class ctx_Pipe(object):
 
-  def __init__(self, fd_state, fd):
-    # type: (FdState, int) -> None
+  def __init__(self, fd_state, fd, job_state, pipeline, last_node):
+    # type: (FdState, int, JobState, Pipeline, command_t) -> None
     fd_state.PushStdinFromPipe(fd)
     self.fd_state = fd_state
+    self.job_state = job_state
+
+    if self.job_state.JobControlEnabled():
+      UP_node = last_node
+
+      # QUESTION: Why only set fg_pipeline when do_fork?
+      # do_fork=False ONLY when CommandEvaluator is called through
+      # main_loop.Batch().  Although I suppose main_loop.Batch() can happen in
+      # 'eval', not just in non-interactive 'osh -c'.
```

Comment:
We need this for when the last stage of the pipeline is a brace group (or potentially any of the other cases when [CommandEvaluator._NoForkLast()](
https://github.com/oilshell/oil/blob/master/osh/cmd_eval.py#L1578-L1612)) would unset this flag.

Reply:
I'll add a comment